tools/lsp/preview/drop_location.rs

```diff
@@ -51,6 +51,139 @@ impl TextOffsetAdjustment {
     }
 }
 
+#[derive(Debug, Eq, PartialEq)]
+enum DropMarkDirection {
+    None,
+    N,
+    E,
+    S,
+    W,
+}
+
+impl DropMarkDirection {
+    /// Create `DropMarkSetup` for a 'live' element (accpting the drop)
+    pub fn for_element(
+        element_geometry: &LogicalRect,
+        position: LogicalPoint,
+        element_is_in_layout: ui::LayoutKind,
+    ) -> Self {
+        if Self::element_accepts(element_geometry, position, element_is_in_layout) {
+            DropMarkDirection::None
+        } else {
+            Self::element_direction(element_geometry, position, element_is_in_layout)
+        }
+    }
+
+    fn border_size(dimension: f32) -> f32 {
+        match dimension {
+            d if d >= 16.0 => 4.0,
+            d if d >= 14.0 => 3.0,
+            d if d >= 12.0 => 2.0,
+            d if d >= 6.0 => 1.0,
+            _ => 0.0,
+        }
+    }
+
+    // Does this hit the center of an element that accepty drops?
+    fn element_accepts(
+        element_geometry: &LogicalRect,
+        position: LogicalPoint,
+        element_is_in_layout: ui::LayoutKind,
+    ) -> bool {
+        if !element_geometry.contains(position) {
+            return false;
+        }
+
+        let mut inner = element_geometry.clone();
```

Comment:
I think LogicalRect is copy and you don't need clone();

But btw, i think you can make the math easier with euclid::Box2d (`.to_box2d()`) (and `Box2D::inflate(-bs_w, -bs_h))`